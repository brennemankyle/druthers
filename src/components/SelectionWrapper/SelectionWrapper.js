import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import AppPropTypes from '../../utils/AppPropTypes'

const preventBlur = e => e.preventDefault()

let SelectionWrapper = (props) => {
  const searchRef = useRef(null)

  let onClick = (e) => {
    let isRemove = e.target.classList.contains('remove')
    e.preventDefault()
    if (isRemove) return

    if (!props.areOptionsOpen) {
      searchRef.current.focus()
    }
    else {
      searchRef.current.blur()
    }
  }

  const Search = React.cloneElement(props.Search, {
    ref: searchRef,
    onFocus: props.onFocus,
    onBlur: props.onBlur,
  })
  let Expand = props.svg_Expand

  return <div
    className={props.className}
    onClick={onClick}
    onMouseDown={preventBlur}
    onTouchStart={preventBlur}>
      <div>
        {props.SelectionList}
        {Search}
      </div>
      <span className="divider"></span>
      <Expand /></div>
}

SelectionWrapper.propTypes = {
  onFocus: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,
  areOptionsOpen: PropTypes.bool.isRequired,
  Search: AppPropTypes.element.isRequired,
  SelectionList: PropTypes.oneOfType([AppPropTypes.element.isRequired, PropTypes.bool]),
  svg_Expand: AppPropTypes.element.isRequired,
  ...AppPropTypes.styles,
}

export default SelectionWrapper
