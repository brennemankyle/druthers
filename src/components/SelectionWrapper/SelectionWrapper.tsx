import React, { useRef, MouseEventHandler, ReactElement } from "react";

interface Props {
  onFocus: MouseEventHandler<HTMLInputElement>;
  onBlur: MouseEventHandler<HTMLInputElement>;
  areOptionsOpen: boolean;
  Search: ReactElement;
  SelectionList: ReactElement;
  svg_Expand: ReactElement;
}

const preventBlur = (e) => e.preventDefault();

let SelectionWrapper = (props) => {
  const searchRef = useRef(null);

  let onClick = (e) => {
    e.preventDefault();
    if (e.target.classList.contains("remove")) return; // Don't open options if remove was clicked

    if (!props.areOptionsOpen) {
      searchRef.current.focus(); // Open options
    } else {
      searchRef.current.blur(); // Close options
    }
  };

  const Search = React.cloneElement(props.Search, {
    ref: searchRef,
    onFocus: props.onFocus,
    onBlur: props.onBlur,
  });
  let Expand = props.svg_Expand;

  return (
    <div
      className={props.className}
      onClick={onClick}
      onMouseDown={preventBlur}
      onTouchStart={preventBlur}
    >
      <div>
        {props.SelectionList}
        {Search}
      </div>
      <span className="divider"></span>
      <Expand />
    </div>
  );
};

export default SelectionWrapper;
