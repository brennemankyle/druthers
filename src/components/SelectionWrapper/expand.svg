<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" preserveAspectRatio="xMidYMid meet" viewBox="0 0 16 8">
  <defs>
    <path d="M4 4L0 0L8 0L16 0L12 4L8 8L4 4Z" id="aVbcAhG7I"></path>
  </defs>
  <g>
    <g>
      <use xlink:href="#aVbcAhG7I" opacity="1" fill="#000000" fill-opacity="1"></use>
    </g>
  </g>
</svg>
