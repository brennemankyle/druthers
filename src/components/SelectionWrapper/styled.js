import styled from "@emotion/styled";
import RawSelectionWrapper from "./SelectionWrapper";

const SelectionWrapper = styled(RawSelectionWrapper)`
  border: 1px solid ${props => props.styles_colors_secondary};
  border-radius: ${props => props.styles_borderRadius};
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-top: ${props => props.styles_paddingTop};
  padding-bottom: ${props => props.styles_paddingBottom};
  padding-right: ${props => props.styles_paddingRight};
  padding-left: ${props => props.styles_paddingLeft};
  & > div:first-of-type {
    flex-grow: 1;
    flex-shrink: 1;
  }
  & > * {
    flex-shrink: 0;
  }

  svg.expand {
    fill: ${props => props.styles_colors_primary};
  }

  ${props =>
    !props.styles_disabled &&
    `&:hover {
      border-color: ${props.styles_colors_primary};
    }`}

  ${props =>
    props.areOptionsOpen &&
    `svg.expand {
      transform: scaleY(-1);
    }
    border-color: ${props.styles_colors_primary};`}

  ${props =>
    props.styles_hasOptions
      ? `.divider {
        width: 1px;
        align-self: stretch;
        background-color: ${props.styles_colors_secondary};
        margin-top: calc(.2em - ${props.styles_paddingTop});
        margin-bottom: calc(.2em - ${props.styles_paddingBottom});
        margin-left: .4em;
        margin-right: .4em;
      }

      svg.expand {
        width: ${props.styles_icon_width};
        height: ${props.styles_icon_width};
        margin-top: -1px;

        ${!props.styles_disabled &&
          `&:hover {
            opacity: 0.5;
          }`}
      }`
      : `.divider {
        display: none;
      }

      .expand {
        display: none;
      }`}

  background-color: ${props =>
    props.styles_disabled
      ? props.styles_colors_disabled
      : props.styles_colors_background};
  ${props => props.styles_disabled && `cursor: not-allowed;`}
`;

export { SelectionWrapper };
