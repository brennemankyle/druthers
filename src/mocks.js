let mockStyles = {
  colors: {}
}

let mockItem = {
  value: '42',
  label: 'Meaning of life',
}

let mockItemList = [
  {value: '1', label: 'Item 1'},
  {value: '2', label: 'Item 2'},
  {value: '3', label: 'Item 3'},
  {value: '4', label: 'Item 4'},
  {value: '5', label: 'Item 5'},
]

export { mockStyles, mockItem, mockItemList }
