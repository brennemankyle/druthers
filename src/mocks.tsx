import React from "react";
import defaultProps from "./utils/defaultProps";
import { withKeys, noop } from "./utils/utils";

let mockStyles = {
  styles_width: 600,
  styles_multiple: false,
  styles_disabled: false,
  styles_hasSelection: false,
  styles_hasOptions: true,
  styles_optionHighlighted: [],
  styles_rightToLeft: false,
  ...withKeys(defaultProps, "styles_"),
};

let mockItem = {
  value: "42",
  label: "Meaning of life",
};

let mockItemList = [
  { value: "1", label: "Item 1" },
  { value: "2", label: "Item 2" },
  { value: "3", label: "Item 3" },
  { value: "4", label: "Item 4" },
  { value: "5", label: "Item 5" },
];

let mockEvent = {
  preventDefault: noop,
};

let MockElement = (props: any) => <div data-props={props} />;
let MockInput = (props: any) => <input type="text" data-props={props} />;
let MockSvg = () => (
  <svg>
    <path d="M0 0L5 5Z" />
  </svg>
);

let mockRect = {
  top: 10,
  bottom: 110,
  left: 20,
  right: 220,
  width: 200,
  height: 100,
  x: 15,
  y: 25,
};

export {
  mockStyles,
  mockItem,
  mockItemList,
  mockEvent,
  MockElement,
  MockInput,
  MockSvg,
  mockRect,
};
