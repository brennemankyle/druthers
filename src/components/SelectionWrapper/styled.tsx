import React from "react";
import styled from "@emotion/styled";
import RawSelectionWrapper from "./SelectionWrapper";

const SelectionWrapper = styled(RawSelectionWrapper)(
  (props) => `
  border: 1px solid ${props.styles_colors_secondary};
  border-radius: ${props.styles_borderRadius};
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-top: ${props.styles_paddingTop};
  padding-bottom: ${props.styles_paddingBottom};
  padding-right: ${props.styles_paddingRight};
  padding-left: ${props.styles_paddingLeft};
  & > div:first-of-type {
    flex-grow: 1;
    flex-shrink: 1;
  }
  & > * {
    flex-shrink: 0;
  }

  svg.expand {
    fill: ${props.styles_colors_primary};
  }

  ${
    !props.styles_disabled
      ? `&:hover {
          border-color: ${props.styles_colors_primary};
        }`
      : ``
  }

  ${
    props.areOptionsOpen
      ? `svg.expand {
          transform: scaleY(-1);
        }
        border-color: ${props.styles_colors_primary};`
      : ``
  }

  ${
    props.styles_hasOptions && !props.styles_optionsAlwaysOpen
      ? `.divider {
          width: 1px;
          align-self: stretch;
          background-color: ${props.styles_colors_secondary};
          margin-top: calc(.2em - ${props.styles_paddingTop});
          margin-bottom: calc(.2em - ${props.styles_paddingBottom});
          margin-left: .4em;
          margin-right: .4em;
        }

        svg.expand {
          width: ${props.styles_icon_width};
          height: ${props.styles_icon_width};
          margin-top: -1px;
          ${
            !props.styles_disabled
              ? `&:hover {
                  opacity: 0.5;
                }`
              : ``
          }
        }`
      : `.divider {
        display: none;
      }

      .expand {
        display: none;
      }`
  }

  background-color: ${
    props.styles_disabled
      ? props.styles_colors_disabled
      : props.styles_colors_background
  };
  ${props.styles_disabled ? `cursor: not-allowed;` : ``}
`
);

export { SelectionWrapper };
