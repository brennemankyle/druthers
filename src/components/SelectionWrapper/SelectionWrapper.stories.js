import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { mockStyles } from '../../mocks'
import props from '../../utils/defaultProps'
import { SelectionWrapper, SelectionList, Search, Selection } from '../styledComponents/styledComponents'

let selectionList = <SelectionList
  itemList={props.selection}
  onClick={action('onRemove')}
  removable={!props.disabled && props.removable}
  Item={Selection}
  {...mockStyles} />
let search = <Search
  hide={false}
  placeholder={props.placeholder}
  searchText={''}
  onKeyDown={action('onKeyDown')}
  onChange={action('onChange')}
  {...mockStyles} />

storiesOf('(Internal) SelectionWrapper', module)
  .add('closed', () => <SelectionWrapper
    onFocus={action('onFocus')}
    onBlur={action('onBlur')}
    {...mockStyles}
    areOptionsOpen={false}
    SelectionList={selectionList}
    Search={search} />)

storiesOf('(Internal) SelectionWrapper', module)
  .add('open', () => <SelectionWrapper
    onFocus={action('onFocus')}
    onBlur={action('onBlur')}
    {...mockStyles}
    areOptionsOpen={true}
    SelectionList={selectionList}
    Search={search} />)

let styles = {
  ...mockStyles,
  styles_disabled: true,
}

storiesOf('(Internal) SelectionWrapper', module)
  .add('disabled', () => <SelectionWrapper
    onFocus={action('onFocus')}
    onBlur={action('onBlur')}
    {...styles}
    areOptionsOpen={false}
    SelectionList={<SelectionList
      itemList={props.selection}
      onClick={action('onRemove')}
      removable={!props.disabled && props.removable}
      Item={Selection}
      {...styles} />}
    Search={<Search
      placeholder={props.placeholder}
      searchText={''}
      onKeyDown={action('onKeyDown')}
      onChange={action('onChange')}
      {...styles} />} />)
