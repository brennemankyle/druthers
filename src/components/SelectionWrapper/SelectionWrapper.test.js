import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { mockStyles, mockEvent, MockElement, MockInput } from '../../mocks'
import SelectionWrapper from './SelectionWrapper'

let SelectionList = (props) => <div {...props} />
let Input = (props, ref) => <input {...props} ref={ref} />
Input = React.forwardRef(Input)
let onFocus = jest.fn()
let onBlur = jest.fn()

it('renders', () => {
  const wrapper = shallow(<SelectionWrapper
    {...mockStyles}
    className="test"
    areOptionsOpen={false}
    onFocus={onFocus}
    onBlur={onBlur}
    SelectionList={MockElement}
    Search={MockInput} />)

  expect(toJson(wrapper)).toMatchSnapshot()
})

it('renders disabled', () => {
  let styles = {
    ...mockStyles,
    styles_disabled: true,
  }

  const wrapper = shallow(<SelectionWrapper
    {...styles}
    className="test"
    areOptionsOpen={false}
    onFocus={onFocus}
    onBlur={onBlur}
    SelectionList={MockElement}
    Search={MockInput} />)

  expect(toJson(wrapper)).toMatchSnapshot()
})

// ref searchRef.current.blur() doesn't continue event propogation
// it('should close', (done) => {
//   const wrapper = mount(<SelectionWrapper
//     {...mockStyles}
//     className="test"
//     areOptionsOpen={true}
//     onFocus={onFocus}
//     onBlur={onBlur}
//     SelectionList={<SelectionList />}
//     Search={<Input />} />)
//
//   setTimeout(() => {
//     wrapper.find('input').simulate('click')
//     expect(onBlur).toBeCalled()
//     done()
//   })
// })
