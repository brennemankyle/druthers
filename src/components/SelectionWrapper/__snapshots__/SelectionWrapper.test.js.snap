// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="test"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="-1"
>
  <div>
    <Component />
    <Component
      onClick={[Function]}
    />
  </div>
  <span
    className="divider"
  />
  <img
    alt="expand"
    className="expand"
    src="expand.png"
  />
</div>
`;

exports[`renders disabled 1`] = `
<div
  className="test"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="false"
>
  <div>
    <Component />
    <Component
      onClick={[Function]}
    />
  </div>
  <span
    className="divider"
  />
  <img
    alt="expand"
    className="expand"
    src="expand.png"
  />
</div>
`;
